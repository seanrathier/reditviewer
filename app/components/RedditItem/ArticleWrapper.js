import styled from 'styled-components';

const ArticleWrapper = styled.article`
  display: flex;
  flex-direction: column;
  margin-bottom: 10px;
  padding: 8px;
  width: 100%;
  background-color: white;
  border: 1px solid #ccc;
  border-radius: 3px;
  overflow: hidden;
`;

export default ArticleWrapper;
