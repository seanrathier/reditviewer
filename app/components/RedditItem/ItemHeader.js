import styled from 'styled-components';

export const ItemHeader = styled.div`
  display: flex;
  justify-content: space-between;
  color: rgb(120, 124, 126);
  font-size: 12px;
  font-weight: 400;
  line-height: 16px;
`;
