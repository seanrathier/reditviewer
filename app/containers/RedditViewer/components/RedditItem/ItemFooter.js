import styled from 'styled-components';

export const ItemFooter = styled.div`
  display: flex;
  justify-content: space-between;
  color: rgb(120, 124, 126);
  font-weight: 700;
  line-height: 16px;
  padding-top: 10px;
`;
