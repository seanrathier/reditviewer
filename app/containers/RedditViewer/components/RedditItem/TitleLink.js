import styled from 'styled-components';

export const TitleLink = styled.a`
  color: #222222;
  text-decoration: none;
`;
