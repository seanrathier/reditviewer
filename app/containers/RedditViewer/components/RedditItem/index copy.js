/**
 * A link to a certain page, an anchor tag
 */

import styled from 'styled-components';

const A = styled.a`
  color: #222222;
  text-decoration: none;

  &:hover {
    font-weight: 600;
  }
`;

export default A;
