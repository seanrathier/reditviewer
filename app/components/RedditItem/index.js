/**
 *
 * RedditItem
 *
 */

import React, { memo } from 'react';
import PropTypes from 'prop-types';
import moment from 'moment-mini';
import ArticleWrapper from './ArticleWrapper';
import { ItemHeader } from './ItemHeader';
import { ItemFooter } from './ItemFooter';
import H3 from '../H3';
import A from '../A';

function RedditItem(props) {
  const { reddit } = props;

  const shortUrl = url => new URL(url).hostname.replace('www.', '');
  const dateFromNow = moment.unix(reddit.created).fromNow();

  return (
    <ArticleWrapper>
      <ItemHeader>
        <div>Posted by {reddit.author}</div>
        <div>{dateFromNow}</div>
      </ItemHeader>
      <H3>{reddit.title}</H3>
      {reddit.url && (
        <div>
          <A href={reddit.url}>{shortUrl(reddit.url)}</A>
        </div>
      )}
      <div>
        {reddit.commentCount && (
          <ItemFooter>
            <A href={`https://www.reddit.com/${reddit.permalink}`}>{reddit.commentCount} Comments</A>
          </ItemFooter>
        )}
        {!reddit.commentCount && (
          <ItemFooter>
            <A href={`https://www.reddit.com/${reddit.permalink}`}>Comment</A>
          </ItemFooter>
        )}
      </div>
    </ArticleWrapper>
  );
}

RedditItem.propTypes = {
  reddit: PropTypes.object.isRequired,
};

export default memo(RedditItem);
