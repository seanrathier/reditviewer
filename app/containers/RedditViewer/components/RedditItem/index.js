/**
 *
 * RedditItem
 *
 * This is the component that renders the individual Reddit posts.
 * It accepts a property which is an individual Reddit post.
 * 1. It shortens the url using regex to the domain of the link.
 * 2. Uses moment-mini to get the time from the utc date the post was created
 * 3. Figures out if 'Comment' or 'Comments' should be used fot the comment link
 * 4. Adds the number of comments with a link to the Reddit post
 * 5. Shows the post title that links to the Reddit post (permalink)
 * 6. Shows the thumnai if there is one present.
 *
 */

import React, { memo } from 'react';
import PropTypes from 'prop-types';
import moment from 'moment-mini';
import ArticleWrapper from './ArticleWrapper';
import { ItemHeader } from './ItemHeader';
import { ItemFooter, CommentLink } from './ItemFooter';
import { TitleLink } from './TitleLink';
import Img from '../../../../components/Img';
import H3 from '../../../../components/H3';
import A from '../../../../components/A';

function RedditItem(props) {
  const { reddit } = props;

  const shortUrl = url => new URL(url).hostname.replace('www.', '');
  const dateFromNow = moment.unix(reddit.created).fromNow();
  const commentLink =
    reddit.commentCount && reddit.commentCount > 0
      ? `${reddit.commentCount} Comments`
      : `Comment`;
  const permaLink = `https://www.reddit.com/${reddit.permalink}`;

  return (
    <ArticleWrapper>
      <ItemHeader>
        <div>Posted by {reddit.author}</div>
        <div>{dateFromNow}</div>
      </ItemHeader>
      <H3>
        <TitleLink href={permaLink}>{reddit.title}</TitleLink>
      </H3>
      {reddit.url && (
        <div>
          <A href={reddit.url}>{shortUrl(reddit.url)}</A>
        </div>
      )}
      {reddit.thumbnail && reddit.thumbnail !== 'self' && (
        <div>
          <Img
            alt={reddit.title}
            src={reddit.thumbnail}
            height={reddit.thumbnailHeight}
            width={reddit.thumbnailWidth}
          />
        </div>
      )}
      <ItemFooter>
        <CommentLink href={permaLink}>{commentLink}</CommentLink>
      </ItemFooter>
    </ArticleWrapper>
  );
}

RedditItem.propTypes = {
  reddit: PropTypes.object.isRequired,
};

export default memo(RedditItem);
