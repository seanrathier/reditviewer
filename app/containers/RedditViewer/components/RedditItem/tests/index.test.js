/**
 *
 * Tests for RedditItem
 *
 * @see https://github.com/react-boilerplate/react-boilerplate/tree/master/docs/testing
 *
 */

import React from 'react';
import { render } from 'react-testing-library';
// import 'jest-dom/extend-expect'; // add some helpful assertions

import RedditItem from '../index';

const mockReddit = {
  id: 'fakeId',
  title: 'fakeTitle',
  author: 'fakeAuthor',
  url: 'http://fakeUrl.com/somethingelse',
  created: 152334665,
  score: 23,
  permalink: 'fake/path/to/reddit/2134234',
  commentCount: 23,
  thumbnail: 'http://reddit.com/path/to/thumbnail.jpg',
  thumbnailWidth: 100,
  thumbnailHieght: 100,
};

describe('<RedditItem />', () => {
  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    render(<RedditItem reddit={mockReddit} />);
    expect(spy).not.toHaveBeenCalled();
  });

  it('Expect Comments to be 23 Comments', () => {
    const { getByText } = render(<RedditItem reddit={mockReddit} />);
    expect(getByText('23 Comments')).toBeTruthy();
  });

  it('Expect Title to be exist', () => {
    const { getByText } = render(<RedditItem reddit={mockReddit} />);
    expect(getByText('fakeTitle')).toBeTruthy();
  });

  /**
   * @see {@link https://jestjs.io/docs/en/api#testskipname-fn}
   */
  it('Should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(<RedditItem reddit={mockReddit} />);
    expect(firstChild).toMatchSnapshot();
  });
});
