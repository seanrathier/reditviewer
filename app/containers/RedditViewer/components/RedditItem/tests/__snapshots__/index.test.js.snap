// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RedditItem /> Should render and match the snapshot 1`] = `
<article
  class="ArticleWrapper-sc-17yr5sl-0 izUUIe"
>
  <div
    class="ItemHeader-dmv6uv-0 kVQhFP"
  >
    <div>
      Posted by 
      fakeAuthor
    </div>
    <div>
      45 years ago
    </div>
  </div>
  <h3>
    <a
      class="TitleLink-kxvlag-0 kSfxta"
      href="https://www.reddit.com/fake/path/to/reddit/2134234"
    >
      fakeTitle
    </a>
  </h3>
  <div>
    <a
      class="A-br8h0y-0 kwPCu"
      href="http://fakeUrl.com/somethingelse"
    >
      fakeurl.com
    </a>
  </div>
  <div
    class="ItemFooter-dln1x-0 iRenms"
  >
    <img
      alt="fakeTitle"
      src="http://reddit.com/path/to/thumbnail.jpg"
      width="100"
    />
  </div>
  <div
    class="ItemFooter-dln1x-0 iRenms"
  >
    <a
      class="A-br8h0y-0 kwPCu"
      href="https://www.reddit.com/fake/path/to/reddit/2134234"
    >
      23 Comments
    </a>
  </div>
</article>
`;
